File: scripts/simulate_carrier.py
import requests
import time
import math
import random

# Configuration
API_URL = "http://localhost:8000/carrier/live-location"
CARRIER_ID = "fb1229db-3774-4619-89c8-7779138f3932"

# Start coordinates (Chennai area or as per your map settings)
START_LAT = 13.0827
START_LNG = 80.2707

def simulate_movement():
    print(f"Starting simulation for carrier: {CARRIER_ID}")
    print(f"Target URL: {API_URL}")
    
    step = 0
    lat = START_LAT
    lng = START_LNG
    
    while True:
        try:
            # Simulate a small movement (approx 20-40 meters)
            # 0.0001 degrees is roughly 11 meters
            lat_change = (math.sin(step * 0.1) * 0.0002) + 0.0001
            lng_change = (math.cos(step * 0.1) * 0.0002) + 0.0001
            
            lat += lat_change
            lng += lng_change
            
            # Calculate heading in degrees (0-360)
            heading = (step * 5) % 360
            
            # Simulate realistic speed (40-65 km/h)
            speed = 45 + (random.random() * 10)
            
            payload = {
                "carrier_id": CARRIER_ID,
                "lat": lat,
                "lng": lng,
                "speed": speed,
                "heading": heading
            }
            
            response = requests.post(API_URL, json=payload, timeout=5)
            
            if response.status_code == 200:
                print(f"[{time.strftime('%H:%M:%S')}] Step {step}: Updated -> Lat: {lat:.6f}, Lng: {lng:.6f}, Speed: {speed:.1f} km/h")
            else:
                print(f"[{time.strftime('%H:%M:%S')}] Failed to update. Status: {response.status_code}")
                
        except requests.exceptions.ConnectionError:
            print(f"[{time.strftime('%H:%M:%S')}] Connection Error: Is the server.py running on port 8000?")
        except Exception as e:
            print(f"[{time.strftime('%H:%M:%S')}] Error: {str(e)}")
            
        step += 1
        time.sleep(1) # Send update every second

if __name__ == "__main__":
    simulate_movement()

File: scripts/simulate_carrier.js
const fetch = require('node-fetch');

async function simulate() {
    const carrierId = process.argv[2] || "fb1229db-3774-4619-89c8-7779138f3932";
    console.log(`Starting simulation for carrier: ${carrierId}`);

    // Predefined route for simulation (Chennai area)
    const route = [
        { lat: 13.0827, lng: 80.2707 },
        { lat: 13.0827, lng: 80.2907 }, // Move East
        { lat: 13.1027, lng: 80.2907 }, // Move North
        { lat: 13.1027, lng: 80.2707 }, // Move West
        { lat: 13.0827, lng: 80.2707 }, // Move South back to start
    ];

    let currentWaypoint = 0;
    let lat = route[0].lat;
    let lng = route[0].lng;
    let heading = 0;
    let counter = 0;

    const interval = setInterval(async () => {
        counter++;
        const nextWaypoint = (currentWaypoint + 1) % route.length;
        const target = route[nextWaypoint];

        // Cycle traffic: 15 ticks clear (~45s), 5 ticks congested (~15s)
        const isCongested = (counter % 20 > 15);
        const targetSpeed = isCongested ? 5 + Math.random() * 5 : 45 + Math.random() * 15;

        // Calculate move step
        const stepSize = (targetSpeed / 3600) * 0.05; // 0.05 degrees covers more ground for demo

        // Calculate vector
        const dLat = target.lat - lat;
        const dLng = target.lng - lng;
        const distance = Math.sqrt(dLat * dLat + dLng * dLng);

        if (distance < stepSize) {
            // Reached waypoint
            currentWaypoint = nextWaypoint;
            lat = target.lat;
            lng = target.lng;
        } else {
            // Move towards target
            lat += (dLat / distance) * stepSize;
            lng += (dLng / distance) * stepSize;

            // Update heading (degrees)
            heading = (Math.atan2(dLng, dLat) * 180 / Math.PI + 360) % 360;
        }

        try {
            const response = await fetch('http://localhost:8000/carrier/live-location', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                    carrier_id: carrierId,
                    lat: lat,
                    lng: lng,
                    speed: targetSpeed,
                    heading: heading
                })
            });
            const result = await response.json();
            const statusIcon = isCongested ? '🛑 ROUTE DELAY' : '🟢 ON ROUTE';
            console.log(`[${new Date().toLocaleTimeString()}] ${statusIcon} | Waypoint: ${currentWaypoint} -> ${nextWaypoint} | Speed: ${targetSpeed.toFixed(1)} km/h | Heading: ${heading.toFixed(0)}°`);
        } catch (error) {
            console.error('Update failed. Is the server.py running?', error.message);
        }
    }, 3000);

    process.on('SIGINT', () => {
        clearInterval(interval);
        console.log('\nSimulation stopped.');
        process.exit();
    });
}

simulate();
